FWCore/Framework/interface/GlobalSchedule.h

```diff
@@ -173,72 +136,88 @@ namespace edm {
-                addContextAndPrintException(ost.str().c_str(), ex, cleaningUpAfterException);
-                excpt = std::current_exception();
-              }
-              if (actReg_) {
-                ServiceRegistry::Operate op(weakToken.lock());
-                actReg_->preGlobalEarlyTerminationSignal_(*globalContext, TerminationOrigin::ExceptionFromThisContext);
-              }
-            }
-            if (actReg_) {
-              // Caught exception is propagated via WaitingTaskHolder
-              CMS_SA_ALLOW try {
-                ServiceRegistry::Operate op(weakToken.lock());
-                T::postScheduleSignal(actReg_.get(), globalContext.get());
-              } catch (...) {
-                if (not excpt) {
-                  excpt = std::current_exception();
-                }
-              }
+              // add context information to the exception and print message
+              handleException(globalContext.get(), weakToken, cleaningUpAfterException, excpt);
             }
+            postScheduleSignal<T>(globalContext.get(), weakToken, excpt);
             iHolder.doneWaiting(excpt);
           });
-      unsigned int managerIndex = principal.index();
-      if constexpr (T::branchType_ == InRun) {
-        managerIndex += numberOfConcurrentLumis_;
-      }
-      WorkerManager& workerManager = workerManagers_[managerIndex];
-      workerManager.resetAll();
-
-      ParentContext parentContext(globalContext.get());
-      //make sure the ProductResolvers know about their
-      // workers to allow proper data dependency handling
-      workerManager.setupResolvers(transitionInfo.principal());
 
       //make sure the task doesn't get run until all workers have beens started
       WaitingTaskHolder holdForLoop(*iHolder.group(), doneTask);
-      auto& aw = workerManager.allWorkers();
-      for (Worker* worker : boost::adaptors::reverse(aw)) {
-        worker->doWorkAsync<T>(
-            holdForLoop, transitionInfo, token, StreamID::invalidStreamID(), parentContext, globalContext.get());
+
+      CMS_SA_ALLOW try {
+        preScheduleSignal<T>(globalContext.get(), token);
+
+        unsigned int managerIndex = principal.index();
+        if constexpr (T::branchType_ == InRun) {
+          managerIndex += numberOfConcurrentLumis_;
+        }
+        WorkerManager& workerManager = workerManagers_[managerIndex];
+        workerManager.resetAll();
+
+        ParentContext parentContext(globalContext.get());
+        // make sure the ProductResolvers know about their
+        // workers to allow proper data dependency handling
+        workerManager.setupResolvers(transitionInfo.principal());
+
+        auto& aw = workerManager.allWorkers();
+        for (Worker* worker : boost::adaptors::reverse(aw)) {
+          worker->doWorkAsync<T>(
+              holdForLoop, transitionInfo, token, StreamID::invalidStreamID(), parentContext, globalContext.get());
```

Comment:
Would `worker->doWorkAsync()` throwing an exception to be a bad thing? (I'd imagine it should not throw exception). This loop just reminds me of the one in https://github.com/cms-sw/cmssw/issues/44744#issuecomment-2065349107 and I'm wondering if we should do the additional hardening (e.g. marking `doWorkAsync()` as `noexcept`)?

Reply:
It should be very rare that ```doWorkAsync``` emits an exception. All the functions calls that are user module functions should have their exceptions caught and returned via the WaitingTask. It is technically possible though. There are little memory allocations (we could run out of memory) and the possibility of a code bug, but the doWorkAsync call is in a try-catch block. If it did throw, the exception would be caught and cmsRun would exit gracefully with an exception message. It should not get stuck or fail in some other bad way. As a test, I just inserted ```throw "foo"``` at that point then ran a cmsRun job. It exited gracefully with the correct exception message (the begin ProcessBlock transition is the first that goes through there):
```
----- Begin Fatal Exception 26-Apr-2024 22:31:34 CEST-----------------------
An exception of category 'BadExceptionType' occurred while
   [0] Processing begin ProcessBlock
Exception Message:
A const char* was thrown as an exception.
foo
----- End Fatal Exception -------------------------------------------------
```
and exit code 69. I think that is better than terminating which is what would happen if we added noexcept. Although would agree that in cases where it gets stuck, terminating would be better.

Reply:
Thanks for the test. If the job doesn't end up stuck even if an exception would be thrown, this should be fine.